add xchain-haven

MartyHav opened the pull request:
This PR adds xchain-haven package. 

a few marks: 

-  xchain-util is included as a local package as the npm version, doesn't include enum Chain.Haven yet
https://github.com/haven-protocol-org/xchainjs-lib/blob/261e0371771a0331fd1489d62f872b0d0f90e90a/packages/xchain-haven/package.json#L37

- mnemonicConverter is included as a GitHub package and set to latest commit. It exposes two functions, where you can either convert a bip39 phrase into a haven or monero mnemonic. It was planed to move this package/functionality to xchain-crypto package and I would like to leave this to xchainjs maintainers. 
https://github.com/haven-protocol-org/xchainjs-lib/blob/261e0371771a0331fd1489d62f872b0d0f90e90a/packages/xchain-haven/package.json#L40

- haven-core-js the underlying wallet library/sdk is included as a GitHub package as well for now. I'll create a npm package of it shortly.

Besides of these points, this PR is ready to be reviewed.

veado commented:
Thx for your PR!

>     * xchain-util is included as a local package as the npm version, doesn't include enum Chain.Haven yet
>       https://github.com/haven-protocol-org/xchainjs-lib/blob/261e0371771a0331fd1489d62f872b0d0f90e90a/packages/xchain-haven/package.json#L37
Pls add `Chain.Haven` to `xchain-util`, bump `xchain-util` to next patch version + add this version to `xchain-haven`. To test it locally in your `xchain-haven, just run `yarn build` in `xchain-util` once.

>     * mnemonicConverter is included as a GitHub package and set to latest commit. It exposes two functions, where you can either convert a bip39 phrase into a haven or monero mnemonic. It was planed to move this package/functionality to xchain-crypto package and I would like to leave this to xchainjs maintainers.
>       https://github.com/haven-protocol-org/xchainjs-lib/blob/261e0371771a0331fd1489d62f872b0d0f90e90a/packages/xchain-haven/package.json#L40
Pls add it to `xchain-util` as well (another PR is recommend, this PR ^ has already ~5k LOC).

>     * haven-core-js the underlying wallet library/sdk is included as a GitHub package as well for now. I'll create a npm package of it shortly.
Yes, pls add a npm package.

Thanks again!

mikewyszinski commented:
also can you merge the latest from master and add a changelog?

MartyHav commented:
@veado Implemented all your remarks and added a new PR: https://github.com/xchainjs/xchainjs-lib/pull/593 for bip 39 phrase converter. Just saw you wanted to have it in xchain-util and not in xchain-crypto, would be a quick change, we can sort it out there.

MartyHav commented on packages/xchain-haven/src/client.ts:
```diff
@@ -0,0 +1,320 @@
+// rootDerivationPaths = {
+//   [Network.Mainnet]: `m/44'/60'/0'/0/`,
+//   [Network.Testnet]: `m/44'/60'/0'/0/`, // this is INCORRECT but makes the unit tests pass
+//   [Network.Stagenet]: `m/44'/60'/0'/0/`,
```
done!

MartyHav commented on packages/xchain-haven/package.json:
```diff
@@ -0,0 +1,52 @@
+    "@xchainjs/xchain-crypto": "^0.2.6",
+    "@xchainjs/xchain-util": "file:../xchain-util",
+    "haven-core-js": "haven-protocol-org/haven-core-js#3d8b853f207c307a79b146c6f2b0ce9702f462ae",
+    "mnemonicconverter": "haven-protocol-org/bip39mnemonicConverter#b12a9c0eac4a23b9bcc007ae9494220c5c316556"
```
done!

MartyHav commented on packages/xchain-haven/package.json:
```diff
@@ -0,0 +1,52 @@
+{
+  "name": "@xchainjs/xchain-haven",
+  "version": "0.11.0",
```
done!

MartyHav commented on packages/xchain-haven/__tests__/openhaven-api.test.ts:
```diff
@@ -0,0 +1,56 @@
+
+  // const mnenomonic = `haggled rekindle pimple pebbles dozen zippers perfect olympics
+  // popular frying goblet gemstone snug dads factual loyal
+  // balding annoyed lumber vary welders navy laboratory maverick olympics`
```
done!

MartyHav commented on packages/xchain-haven/README.md:
````diff
@@ -0,0 +1,25 @@
+Following peer dependencies have to be installed into your project. These are not included in `@xchainjs/xchain-haven`.
+
+```
+yarn add @xchainjs/xchain-client @xchainjs/xchain-crypto @xchainjs/xchain-util axios haven-protocol-org/haven-core-js#3d8b853f207c307a79b146c6f2b0ce9702f462ae haven-protocol-org/bip39mnemonicConverter#b12a9c0eac4a23b9bcc007ae9494220c5c316556
````
done!

MartyHav commented on packages/xchain-haven/README.md:
````diff
@@ -0,0 +1,25 @@
+```
+## Documentation
+
+Coming soon...
````
yes, rephrased it as a TODO in readme, will require a PR for xchain docs, which I will do next

MartyHav commented:
> also can you merge the latest from master and add a changelog?

yes merged and added changelog. Should changelog be maintained while PR is ongoing?

veado commented on packages/xchain-haven/src/client.ts:
```diff
@@ -0,0 +1,315 @@
+      assets.forEach((asset) => {
+        const assetBalance: Balance = {
+          asset,
+          amount: baseAmount(havenBalance[asset.ticker as HavenTicker].balance, 12),
```
Type cast is unsafe and might fail. Do a check if `asset.ticker` is really `HavenTicker` by introducing a type guard in `utils`

```
const isHavenTicker = (ticker: string): ticker is HavenTicker => (_...do the check here..._)
```

MartyHav replied:
guard added and in use

veado commented on packages/xchain-haven/src/client.ts:
```diff
@@ -0,0 +1,315 @@
+
+    //filter by date
+    transactions = params?.startTime
+      ? transactions.filter((tx, _) => new Date(tx.timestamp) >= params.startTime!)
```
Lint does not like `!` here `Forbidden non-null assertion`

veado commented on packages/xchain-haven/src/client.ts:
```diff
@@ -0,0 +1,315 @@
+    const amountString = amount.amount().toString()
+    let txHash
+    try {
+      txHash = await this.havenSDK.transfer(amountString, asset.ticker as HavenTicker, recipient, memo)
```
(Similar to a previous comment) Try to use a type guard instead of `as HavenTicker`

veado commented on packages/xchain-haven/src/client.ts:
```diff
@@ -0,0 +1,315 @@
+   * @param observer
+   */
+  subscribeSyncProgress(observer: SyncObserver): void {
+    this.havenSDK.subscribeSyncProgress(observer)
```
Is there a way to unsubscribe an observer if needed (e.g. in `purgeClient`)?

MartyHav replied:
added a unsubscribe method, when `purgeClient()` is called all observers are removed

veado commented on packages/xchain-haven/src/haven/api.ts:
```diff
@@ -0,0 +1,112 @@
+  },
+}
+
+export const get_version = (): Promise<any> => {
```
`Unexpected any. Specify a different type.eslint`

Pls avoid `any` here and anywhere. Try to define API types (similar to other xchain-*` packages)

MartyHav replied:
API Types added

veado commented on packages/xchain-haven/src/haven/haven-core-client.ts:
```diff
@@ -0,0 +1,319 @@
+  }
+
+  async init(seed: string, netType: string | number): Promise<boolean> {
+    //this.netTypeId = netTypePromise<boolean> {
```
Remove ^

veado commented on packages/xchain-haven/src/haven/haven-core-client.ts:
```diff
@@ -0,0 +1,319 @@
+    this.syncHandler.purge()
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    //@ts-ignore
+    clearInterval(this.pingServerIntervalID)
```
Try to follow eslint, e.g.
```
this.pingServerIntervalID && clearInterval(this.pingServerIntervalID)
```

veado commented on packages/xchain-haven/src/haven/haven-core-client.ts:
```diff
@@ -0,0 +1,319 @@
+
+      const lockedBalance = balance.subtract(unlockedBalance)
+
+      havenBalance[assetType as HavenTicker] = {
```
Try to avoid `assetType as HavenTicker` (see previous comments to similar topic)

veado commented on packages/xchain-haven/src/haven/haven-core-client.ts:
```diff
@@ -0,0 +1,319 @@
+
+    if (this.base_fee === undefined || this.fork_version === undefined) {
+      const version = await get_version()
+      this.fork_version = version.fork_version as number
```
No need to cast the result if `get_version` will return a proper type (not `any`). Pls check previous comment to missing API types.

veado commented on packages/xchain-haven/src/haven/haven-core-client.ts:
```diff
@@ -0,0 +1,319 @@
+ * callback functions for sending transfers procedure, used by haven-core-client.ts
+ */
+
+const updateStatus = (_status: any) => {
+  //console.log(status)
+}
```
Is `updateStatus` still needed ^?

veado commented on packages/xchain-haven/src/haven/sync-handler.ts:
```diff
@@ -0,0 +1,95 @@
+    if (!(await this.isSyncing())) {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      //@ts-ignore
+      clearInterval(this.updateSyncProgressIntervalID) // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
No need to ignore/disable eslint: `this.updateSyncProgressIntervalID` might be undefined, just add a check before (similar you have done in `purge`)

veado reviewed (changes requested):
Nice PR! Some comments have been added...

```bash
yarn lint
...
✖ 29 problems (1 error, 28 warnings)
```
Pls don't forget to run `yarn lint`

Thorian1te reviewed (commented):
Looks good to me. 👍

MartyHav commented on packages/xchain-haven/src/client.ts:
```diff
@@ -0,0 +1,315 @@
+    super(Chain.Haven, { network, rootDerivationPaths, phrase })
+    this.havenSDK = new HavenCoreClient()
+    if (this.phrase) {
+      this.initSDK()
```
yes makes it a lot cleaner and failsafe, so a static init function is added and constructor is set to private. 
Be aware that I needed to disallow `setPhrase()` and `setNetwork()` as these functions would technically create a new account which would need a reinitiliaization of the sdk ( login on backend server etc... ). They are throwing an error now with a referring message

MartyHav commented:
> Nice PR! Some comments have been added...
> 
> ```shell
> yarn lint
> ...
> ✖ 29 problems (1 error, 28 warnings)
> ```
> 
> Pls don't forget to run `yarn lint`

Fixed all lint warnings/errors. Also fixed/improved the code parts you addressed in your comments. 
Thanks for the review so far!

MartyHav commented on packages/xchain-haven/src/client.ts:
```diff
@@ -0,0 +1,315 @@
+  }
+
+  async getTransactions(params?: TxHistoryParams): Promise<TxsPage> {
+    const asset: Asset | undefined = params?.asset ? assetFromString(params.asset)! : undefined
```
But this PR would be a breaking change, right?

matikucharski commented:
Guys @veado @mikewyszinski . How is the review going?

parhelium commented:
@veado @mikewyszinski any progress ?

veado commented:
> Guys @veado @mikewyszinski . How is the review going?
> @veado @mikewyszinski any progress ?

@matikucharski @parhelium Sorry for the late reply! Can you or any other Haven team member join THORChain Discord (channel #xchainjs) https://discord.com/invite/pHcS67yX7Z to discuss current status of Haven integration in TC and especially next steps for xchain-haven? Thank you! 

/cc @mikewyszinski @Thorian1te
